File: nest-ts-bun-backend/src/routes/metrics.route.ts
import { Controller, Get, Header } from '@nestjs/common';
import { env } from '../config/env';
import { NotFoundError } from '../errors/http-errors';
import { getMetricsSnapshot, metricsRegistry } from '../observability/metrics';

@Controller('metrics')
export class MetricsController {
  @Get()
  @Header('Content-Type', metricsRegistry.contentType)
  async getMetrics(): Promise<string> {
    if (!env.METRICS_ENABLED) {
      throw new NotFoundError({
        detail: 'Metrics endpoint is disabled.',
      });
    }

    return getMetricsSnapshot();
  }
}

File: flask-python-backend/src/routes/metrics_route.py
from flask import Blueprint, Response

from ..config.env import settings
from ..errors.http_errors import NotFoundError
from ..observability.metrics import get_metrics_snapshot, metrics_content_type

metrics_blueprint = Blueprint('metrics', __name__)


@metrics_blueprint.get('/metrics')
def get_metrics() -> Response:
    if not settings.METRICS_ENABLED:
        raise NotFoundError(detail='Metrics endpoint is disabled.')

    return Response(get_metrics_snapshot(), content_type=metrics_content_type)
